### PR TITLE
Fix #16929 do not show profile load time msg with new pwsh parameter

### LINKED ISSUE #16929: Consider adjusting threshold for showing the load time for personal & system profiles to 1 sec
### Summary of the new feature / enhancement

My profile takes ~640 ms to load which is barely noticeable but I get this extra message every single time I start a new session:

image

With the new minimized startup banner in preview.2, can we consider bumping the threshold for showing the timing info on how long it takes to process my profile?  Maybe from the current 500 ms to 1 s?  It's handy to know if my long startup time is due to my profile but maybe not every time I start a new session.  I know my profile takes a bit longer and I'm at peace with that.  If only PowerShell wouldn't keep reminding me every time it starts.  :-)

Also, I notice that the wall time is noticeably longer than 637ms from seeing `PowerShell 7.xx` to my prompt appearing.  It appears to me to be closer to 1500ms.  So the profile loading accounts for less than half the actual startup time.  But I get that the only knob the user has to twiddle is what they're doing in their profile.

### Proposed technical implementation details (optional)

**Option 1:** Consider adding a new start up parameter `-TimeProfileLoad` or something like that.  Only when this parameter is specified would you display the profile processing time.  In this case, you could also consider being more granular e.g.:
```
> pwsh -TimeProfileLoad
PowerShell 7.3.0-preview.3
 1100ms : Loading C:\Program Files\PowerShell\7-preview\profile.ps1
  637ms : Loading C:\Users\Keith\Documents\PowerShell\profile.ps1
   75ms : Loading C:\Users\Keith\Documents\PowerShell\Microsoft.PowerShell_profile.ps1
 1812ms : Loading system and personal profiles
 C:\Users\Keith> 
 ```

**Option 2:** Perhaps we just bump the current threshold for displaying the message up to 1 sec.

**Option 3:**  We make the threshold configurable.  It defaults to the current 500 ms but folks could adjust that based on their profile.  IOW I might want to see this if it is taking longer than normal to process my profile.  Where "normal" is what you'd configure for your threshold value.

### PR DESCRIPTION
**Pull request by rkeithhill**
This PR adds a new pwsh command line parameter to enable the display of the profile load time message.

# PR Summary

Changes the host to not display the profile load time by default.  A new pwsh command line switch `-ShowProfileLoadTime` was added to allow someone to show the profile load time regardless of the length of the load time.

```
Usage: pwsh[.exe] [-Login] [[-File] <filePath> [args]]
                  [-Command { - | <script-block> [-args <arg-array>]
                                | <string> [<CommandParameters>] } ]
                  [-ConfigurationName <string>] [-CustomPipeName <string>]
                  [-EncodedCommand <Base64EncodedCommand>]
                  [-ExecutionPolicy <ExecutionPolicy>] [-InputFormat {Text | XML}]
                  [-Interactive] [-MTA] [-NoExit] [-NoLogo] [-NonInteractive] [-NoProfile]
                  [-OutputFormat {Text | XML}] [-SettingsFile <filePath>] [-ShowProfileLoadTime]
                  [-SSHServerMode] [-STA] [-Version] [-WindowStyle <style>]
                  [-WorkingDirectory <directoryPath>]

       pwsh[.exe] -h | -Help | -? | /?
       
...

-ShowProfileLoadTime

    Displays the time taken to load the PowerShell profiles.

-SSHServerMode | -sshs
```

## PR Context

Addresses #16929 to further reduce the amount of default startup banner text.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed:  I wasn't going to file an issue on https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_pwsh?view=powershell-7.2 until the PR gets a thumbs up.
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by SteveL-MSFT** [comment]
I believe the original intent of showing the load time was all the complaints from users about how long it was taking PowerShell to startup when most of the time was in their profile.  If this is not enabled by default, I can't imagine anyone taking the time to enable this.  I wonder if it makes more sense to have this opt-out rather than opt-in?  `-DisableProfileLoadTime`.

**Comment by rkeithhill** [comment]
FYI I have re-implemented this as an opt-out parameter `-DisableProfileLoadTime`.  Unfortunately, this approach does not lend itself to unit testing.

**Comment by daxian-dbw** [comment]
@rkeithhill The Console-Interactive WG reviewed this PR and we agree to have it opt-out rather than opt-in. So, by default, the existing behavior is not changed.

Then the ask becomes to have a new switch to hide the profile load time. However, the `-nologo` switch already hides the profile load time, which makes having a new switch just for that seem to be an overkill ...
@rkeithhill what's your thoughts about this?

**Comment by rkeithhill** [comment]
The difference with `-NoLogo` is that you don't see anything displayed in the console until after all profiles are processed.  Where you typically want this feature is when your profile always takes several seconds to process and seeing the profile load time message every single session start is just, well, annoying.  Also, I prefer seeing which version of PowerShell I'm running which you lose with `-NoLogo`.   My preference would be to add an opt-out parameter.  I have it currently implemented with @SteveL-MSFT's suggested name `-DisableProfileLoadTime` but I wonder if `-NoProfileLoadTime` would be better (shorter and more consistent with `-NoLogo`)?

**Comment by SteveL-MSFT** [comment]
I agree that `-NoLogo` suppresses everything whereas you may want the PowerShell version.  `-NoProfileLoadTime` seems fine to me.

**Comment by daxian-dbw** [comment]
Alright, @rkeithhill can you update this PR then? It's still the `showprofileloadtime` implementation.

**Comment by rkeithhill** [comment]
One unfortunate consequence of using an opt-out parameter is that it doesn't lend itself to testing.  From what I can see, the tests in ConsoleHost.Tests.ps1 all specify starting pwsh with `-NoProfile`. That makes sense.  You wouldn't want to process a dev's profile when running console host tests.

**Comment by rkeithhill** [comment]
BTW I did update this PR to be `opt-out`.  It's ready for review again.

**Review by iSazonov — APPROVED** [review]

**Review by PaulHigin — APPROVED** [review]

**Comment by rkeithhill** [comment]
Here's the updated usage:
```
Usage: pwsh[.exe] [-Login] [[-File] <filePath> [args]]
                  [-Command { - | <script-block> [-args <arg-array>]
                                | <string> [<CommandParameters>] } ]
                  [-ConfigurationName <string>] [-CustomPipeName <string>]
                  [-EncodedCommand <Base64EncodedCommand>]
                  [-ExecutionPolicy <ExecutionPolicy>] [-InputFormat {Text | XML}]
                  [-Interactive] [-MTA] [-NoExit] [-NoLogo] [-NonInteractive] [-NoProfile]
                  [-NoProfileLoadTime] [-OutputFormat {Text | XML}]
                  [-SettingsFile <filePath>] [-SSHServerMode] [-STA]
                  [-Version] [-WindowStyle <style>]
                  [-WorkingDirectory <directoryPath>]

       pwsh[.exe] -h | -Help | -? | /?
  
  ...
  
  -NoLogo | -nol

    Hides the banner text at startup of interactive sessions.

...

-NoProfile | -nop

    Does not load the PowerShell profiles.

-NoProfileLoadTime

    Hides the PowerShell profile load time text shown at startup when the load
    time exceeds 500 milliseconds.
```

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs [hunk 1]
```diff
@@ -920,6 +931,11 @@ private void ParseHelper(string[] args)
                     _sshServerMode = true;
                     ParametersUsed |= ParameterBitmap.SSHServerMode;
                 }
+                else if (MatchSwitch(switchKey, "noprofileloadtime", "noprofileloadtime"))
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:934]: Should we consider having `-nplt`?

**Review by SteveL-MSFT — APPROVED** [review]
One non-blocking suggestion

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs (hunk 1 above)
Comment (rkeithhill) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:934]: I like that.  But how about `-noplt` so the alias is a bit more consistent with the other `-no` aliases: `-nol, -nop, -noni`?

Comment (rkeithhill) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:934]: Nevermind, when I try that I run smack into this dbg assert:
```
Dbg.Assert(match.Contains(smallestUnambiguousMatch), "sUM should be a substring of match");
```

**Pull request merged**